mobile/src/auth/jwt-token-handler.ts

```diff
@@ -0,0 +1,114 @@
+import { AccessCodeRequest } from './access-code-request';
+import moment from 'moment';
+import { map } from 'rxjs/operators';
+import { RefreshTokenStorage } from './refresh-token-storage';
+import { Observable, from, Subject, BehaviorSubject } from 'rxjs';
+import { OauthError } from './oauth-error';
+import { RefreshToken } from './login-request';
+
+export class NoJwtTokenError extends Error {
+    constructor() {
+        super('No JWT token is set, user is unauthenticated');
+    }
+}
+
+export class JwtToken {
+    private static readonly clockScrew = moment.duration(1, 'minute');
+
+    constructor(
+        public readonly validUntil: moment.Moment,
+        public readonly value: string
+    ) { }
+
+    public get isExpired() {
+        return moment().isSameOrAfter(this.validUntil);
+    }
+
+    public get isExpiredOrAboutTo() {
```

Comment:
needsRefresh?

Reply:
not his responsibility. it won't refresh itself.

Reply:
isExpired? It have exactly this meaning for the class user.

Reply:
`isExpired` says something objective, its about state. 'needsRefresh' is not objective, its subjective and means something for class user only. Why would token need refresh if its state never changes? The only reason it might need refresh if it knows that expired tokens, or tokens that are about to expire are not useful for HTTP requests; moreover, they should be thrown away. Encapsulation violation :P

Reply:
I mean, why not to remove current isExpired and rename isExpiredOrAboutTo to just isExpired?

Reply:
because it's not actually Expired :) I can remove isExpired as it's not used... I don't know really, I was thinking about this piece myself :)